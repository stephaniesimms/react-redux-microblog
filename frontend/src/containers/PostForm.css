.PostForm { 
  border: 2px solid rgba(0,0,0,.125); 
  border-radius: 10px; 
  padding: 10px; 
  margin: auto;
  max-width: 700px;
}