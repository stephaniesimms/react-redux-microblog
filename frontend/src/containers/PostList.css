.PostList {
  margin: auto;
  max-width: 90%;
  align-items: center;
}

.PostList-card {
  display: inline-block;
  width: 40vw;
  max-width: 50%;
  margin-right: 1em;
}

.PostList-card-title {
  padding: 1em 0em 1em 0em;
}