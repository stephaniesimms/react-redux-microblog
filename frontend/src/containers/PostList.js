import React, { Component } from 'react';
// import PostTitle from '../components/PostTitle';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Card from 'react-bootstrap/Card';
import { getTitlesFromAPI, sendVoteToAPI } from '../actions';

// FIXME: need to rerender PostTitle and sendVoteToAPI from there when user clicks vote
// add votes and display posts by popularity

/** Show list of blog titles, ordered by popularity. */
class PostList extends Component {

  async componentDidMount() {
    if (this.props.titles.length === 0) {
      await this.props.getTitlesFromAPI();
    }
  }
  
  vote(direction, id) {
    console.log("sending vote to API", id, direction)
    this.props.sendVoteToAPI(id, direction);
  }

  render() {
    if (!this.props.titles) return <b>Loading...</b>;
    console.log(this.props)

    // const titles = this.props.titles.map(
    //   (title) =>
    //     <PostTitle key={title.id}
    //       id={title.id}
    //       title={title.title}
    //       description={title.description}
    //       votes={title.votes}
    //       doVote={this.vote} />
    // );

    const titleList = this.props.titles.map(
      (title) =>
        <Card key={title.id} id={title.id}
          style={{ width: '40vw', display: 'inline-block', marginRight: '15px', maxWidth: '300px' }}>
          <Link to={`/posts/${title.id}`}>
            <Card.Title style={{ marginTop: '15px' }}>{title.title}</Card.Title>
          </Link>

          <Card.Body>{title.description}</Card.Body>
          <div className="card-footer">
            <small>{title.votes} votes</small>
            <i className="fas fa-thumbs-up text-success ml-2"
              onClick={() => this.vote("up", title.id)} />
            <i className="fas fa-thumbs-down text-danger ml-2"
              onClick={() => this.vote("down", title.id)} />
          </div>
        </Card>

    );

    return (
      <div>{titleList}</div>
    );
  }
}

function mapStateToProps(state) {
  return { titles: state.titles };
}

export default connect(
  mapStateToProps,
  { getTitlesFromAPI, sendVoteToAPI }
)(PostList);