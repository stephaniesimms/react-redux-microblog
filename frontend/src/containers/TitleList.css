.TitleList-card {
  margin: 1em 0 1em 0;
}

.TitleList-card-title {
  padding: 1em 0 1em 0;
}