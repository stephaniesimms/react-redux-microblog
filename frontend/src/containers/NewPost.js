import React, { Component } from 'react';
import { connect } from 'react-redux';
import PostForm from '../components/PostForm';
import {
  sendPostToAPI,
} from '../actions/actions';

//NOTE: must export connected components for testing
export class NewPost extends Component {
  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.cancel = this.cancel.bind(this);
  }

  /** Handle new post: adds to backend. */

  handleSubmit({ title, description, body }) {
    this.props.sendPostToAPI({ title, description, body });
    this.props.history.push('/');
  }

  /** Cancel (redirect) */

  cancel() {
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h3>New Post</h3>
        <PostForm
          save={this.handleSubmit}
        />
      </div>
    );
  }
}


export default connect(
  null, { sendPostToAPI }
)(NewPost);