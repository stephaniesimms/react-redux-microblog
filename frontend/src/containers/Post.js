import React, { Component } from 'react';
import { connect } from 'react-redux';
import PostForm from '../components/PostForm';
import PostDisplay from '../components/PostDisplay';
import CommentList from '../components/CommentList';
import CommentForm from '../components/CommentForm';
import NotFound from '../components/NotFound';
import {
  sendDeleteToAPI,
  sendUpdateToAPI,
  getPostFromAPI,
  sendVoteToAPI,
  sendCommentToAPI,
  deleteCommentFromAPI
} from '../actions/actions';

/** Post:
 *
 * - get post data from API, if not present
 * - allows post to be edited (showEditForm is local state for this)
 * - handles edit form submission
 * - handles add-comment form submission
 * - handles comment-deletion
 * - handles post-deletion
 */

//NOTE: must export connected components for testing
export class Post extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editing: false
    }

    this.handleDelete = this.handleDelete.bind(this);
    this.handleEdit = this.handleEdit.bind(this);
    this.showEditForm = this.showEditForm.bind(this);
    this.addComment = this.addComment.bind(this);
    this.deleteComment = this.deleteComment.bind(this);
    this.vote = this.vote.bind(this);
  }

  /** If we don't have the post, request it from API. */

  async componentDidMount() {
    if (!this.props.posts) {
      await this.props.getPostFromAPI(this.props.id);
    }
  }

  /** Toggle state of editing true/false to show or hide post edit form */
  showEditForm() {
    this.setState( state => ({ editing: !state.editing }));
  }

  /** Handle post deletion: deletes from backend. */

  async handleDelete() {
    const postId = this.props.match.params.postId;
    await this.props.sendDeleteToAPI(postId);

    this.props.history.push('/');
  }

  /** Handle post editing: adds to backend. */

  handleEdit(updatedPost) {
    const postId = this.props.match.params.postId;
    this.props.sendUpdateToAPI(updatedPost, postId);

    this.showEditForm();
  }

  /** Handle adding a comment: adds to backend. */

  addComment(text) {
    this.props.sendCommentToAPI(this.props.post.id, text);
  }

  /** Handle deleting a comment in backend. */

  deleteComment(commentId) {
    this.props.deleteCommentFromAPI(this.props.post.id, commentId);
  }

  /** Handle voting in backend. */

  vote(direction) {
    this.props.sendVoteToAPI(this.props.post.id, direction)
  }

  /** Render:
   *
   * - if backend does not find post, display 404 not found msg
   * - if no post yet, show a loading message
   * - if editing, show the edit form & comments
   * - if not editing, show the post & comments
   */

  render() {
    if (this.props.error) {
      return <NotFound />;
    }

    const post = this.props.post;

    if (!post) {
      return <b>'Loading...'</b>;
    }

    return (
      <div className='container-fluid'>
  
        {this.state.editing 
          ? <PostForm post={post} 
            id={post.id} 
            save={this.handleEdit}
            />
          : <PostDisplay post={post}
          showEditForm={this.showEditForm}
          deletePost={this.handleDelete}
          doVote={this.vote} />}

        <section c>
          <hr></hr>
          <h4 className="mb-3">Comments</h4>
          <CommentList comments={post.comments}
            deleteComment={this.deleteComment}
            postId={post.id}
          />
          <CommentForm submitCommentForm={this.addComment} />
        </section>
      </div>
    );
  }
}

function mapDispatchToProps(state, props) {
  let id = props.match.params.postId;
  return {
    id,
    post: state.posts[id],
    error: state.error
  };
}

export default connect(
  mapDispatchToProps,
  {
    sendDeleteToAPI,
    sendUpdateToAPI,
    getPostFromAPI,
    sendCommentToAPI,
    deleteCommentFromAPI,
    sendVoteToAPI,
  }
)(Post);