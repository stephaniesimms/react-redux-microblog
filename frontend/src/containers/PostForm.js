import React, { Component } from 'react';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col'
import './PostForm.css';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { sendPostToAPI } from '../actions/actions';

// TODO: should PostForm be a component instead? 
// maybe refactor to follow solution
// otherwise update docstring

/** Show post form.
 *
 * Can be used for editing/adding --- it just calls the parent with edit
 * or cancel actions.
 *
 */
class PostForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      title: '',
      description: '',
      body: '',
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    if (this.props.post) {
      const { title, description, body } = this.props.post;

      this.setState({
        title,
        description,
        body,
      });
    }
  }

  handleChange(evt) {
    this.setState({
      [evt.target.name]: evt.target.value
    });
  };

  /** Checks the formType prop sent from Post component to decide 
   * whether to create a new post or edit an existing post
  */
  async handleSubmit(evt) {
    evt.preventDefault();

    if (this.props.formType === 'Edit') {
      this.props.updatePost({
        ...this.state,
        comments: this.props.post.comments
      });
    } else {
      const { title, description, body } = this.state;
      await this.props.sendPostToAPI({ title, description, body });

      this.props.history.push('/');
    }
  };

  render() {
    return (
      <div className='PostForm'>
        <h3>{this.props.formType}</h3>
        <Form onSubmit={this.handleSubmit}>
          <Form.Group as={Row}>
            <Form.Label column sm={2}>Title</Form.Label>
            <Col sm={10}>
              <Form.Control id='title'
                name='title'
                onChange={this.handleChange}
                value={this.state.title}
                size='md'
                type='text'
                alt='title' />
            </Col>
          </Form.Group>
          <Form.Group as={Row}>
            <Form.Label column sm={2}>Description</Form.Label>
            <Col sm={10}>
              <Form.Control id='description'
                name='description'
                onChange={this.handleChange}
                value={this.state.description}
                size='md'
                type='text'
                alt='description' />
            </Col>
          </Form.Group>
          <Form.Group as={Row}>
            <Form.Label column sm={2}>Body</Form.Label>
            <Col sm={10}>
              <Form.Control id='body'
                name='body'
                onChange={this.handleChange}
                value={this.state.body}
                size='md'
                as='textarea'
                alt='body' rows='5' />
            </Col>
          </Form.Group>
          <div className='row justify-content-end'>
          <Button className='mr-2' type='submit' variant='dark'>
            Post
          </Button>
          <Link to='/'>
              <Button className='mr-3' variant='danger'>
              Cancel
            </Button>
          </Link>
          </div>
        </Form>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return { posts: state.posts };
}


export default connect(mapStateToProps, { sendPostToAPI })(PostForm);