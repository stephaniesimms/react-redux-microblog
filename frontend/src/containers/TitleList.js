import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Card from 'react-bootstrap/Card';
import './TitleList.css';
import {
  getPostsFromAPI,
  sendVoteToAPI
} from '../actions/actions';

/** Show list of blog titles, ordered by popularity.
 *  Allow up- and down-voting in the main list & for individual posts
 */

 //NOTE: must export connected components for testing
export class TitleList extends Component {

  async componentDidMount() {
    await this.props.getPostsFromAPI();
  }

  vote(direction, id) {
    this.props.sendVoteToAPI(id, direction);
  }

 /** Display blog posts by highest to lowest number of votes 
  * NOTE: This method becomes inefficient at scale.
  * One alternative approach is to sort in Redux.
 */
  rankPosts(titleList) {
    return titleList.sort((a, b) => b.votes - a.votes); 
  }

  render() {
    if (Object.keys(this.props.titles).length === 0) {
      return <b>Loading...</b>;
    }

    const sortedTitles = this.rankPosts(this.props.titles);

    const titleList = sortedTitles.map(
      (title) =>
        <div key={title.id} className='col-sm-6'>
          <Card id={title.id} className='TitleList-card'>
            <Card.Body>
              <Link to={`/posts/${title.id}`}>
                <h6 className='TitleList-card-title'>{title.title}</h6>
              </Link>
              <p>{title.description}</p>
            </Card.Body>

            <Card.Footer>
              <small>{title.votes} votes</small>
              <i className='fas fa-grin-stars text-warning ml-2'
                onClick={() => this.vote('up', title.id)} />
              <i className='fas fa-dizzy text-danger ml-2'
                onClick={() => this.vote('down', title.id)} />
            </Card.Footer>
          </Card>
        </div>
    );

    return (
      <div className='row'>
        {titleList}
      </div>
    );
  }
}

function mapStateToProps(state) {
  const titles = Object.entries(state.posts).map(([key, post]) => ({
    id: post.id,
    title: post.title,
    description: post.description,
    votes: post.votes
  }));
  return { titles: titles };
}

export default connect(
  mapStateToProps,
  { getPostsFromAPI, sendVoteToAPI }
)(TitleList);