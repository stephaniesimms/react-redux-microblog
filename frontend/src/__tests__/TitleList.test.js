import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { TitleList } from '../containers/TitleList';

//FIXME: determine correct way to mock titles
/**
 * Without any handling of the Titlelist properties we get the following error:
 * (node:31957) UnhandledPromiseRejectionWarning: TypeError: this.props.getPostsFromAPI is not a function
(node:31957) UnhandledPromiseRejectionWarning: Unhandled promise rejection.
 * 
 * To circumvent this we can do a hack: Feed in a `getPostsFromAPI` to mock setting a function that is
 *  then being called. This works since we disable the lifecycle methods and use a spy to check we call the
 *  lifecycle without actually executing it. However, the clean way to do it is to mock the Redux Store in
 *  order to have the connector set the property as it does in production. 
*/
const TITLES = [{id: 1, title: "test title", description: "test description", votes: 2}]

beforeAll(() => {
  global.fetch = jest.fn();
  //window.fetch = jest.fn(); if running browser environment
});

let wrapper;
beforeEach(() => {
  wrapper = shallow(<TitleList titles={TITLES} getPostsFromAPI={(x) => (x)}/>, { disableLifecycleMethods: true });
});

afterEach(() => {
  wrapper.unmount();
});


it('matches snapshot', function () {
  let serialized = toJson(wrapper);
  expect(serialized).toMatchSnapshot();
});


it('spy whether componentDidMount is called', () => {
  const spyDidMount = jest.spyOn(TitleList.prototype, "componentDidMount");

  fetch.mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      json: () => {
        return Promise.resolve(TITLES);
      }
    });
  });

  wrapper.instance().componentDidMount();
  expect(spyDidMount).toHaveBeenCalled();

});
