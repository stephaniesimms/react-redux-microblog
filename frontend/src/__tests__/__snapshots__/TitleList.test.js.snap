// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-6"
    key="1"
  >
    <Card
      body={false}
      className="TitleList-card"
      id={1}
    >
      <CardBody>
        <Link
          to="/posts/1"
        >
          <h6
            className="TitleList-card-title"
          >
            test title
          </h6>
        </Link>
        <p>
          test description
        </p>
      </CardBody>
      <CardFooter>
        <small>
          2
           votes
        </small>
        <i
          className="fas fa-grin-stars text-warning ml-2"
          onClick={[Function]}
        />
        <i
          className="fas fa-dizzy text-danger ml-2"
          onClick={[Function]}
        />
      </CardFooter>
    </Card>
  </div>
</div>
`;
